src/client/mod.rs

```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
+    pub can_have_body: bool,
+}
+
+/// A trait that represents the functionality of creating a brand new fresh
+/// request (an instance of a type implementing `FreshHttpRequest`).
+///
+/// Each particular implementation of the trait can provide a different way
+/// for achieving this, as well as a custom type of request that it produces.
+///
+/// Therefore, by defining the type of the fresh request that it will produce,
+/// it also directly defines the type of the response, since the chain of
+/// `Fresh -> Streaming -> Response` types is always well defined given a
+/// particular `FreshHttpRequest` starting point.
+///
+/// TODO: Might need a better name?
+pub trait HttpRequestFactory {
+    /// The type of the `FreshHttpRequest` that this factory will produce.
+    type RequestType: FreshHttpRequest;
+    /// Creates a new instance of the particular `HttpRequestFactory`.
+    /// This method is a convenience so that the `Client` can still be
+    /// instantiated without providing a factory instance, but with a simple
+    /// `new()` call. This, however, means that all particular
+    /// `HttpRequestFactory` implementations must have a set of default
+    /// parameters with which they can be created.
+    fn new() -> Self;
+    /// Creates a new `FreshHttpRequest` of the appropriate type given the
+    /// `RequestTemplate` instance.
+    fn get_fresh_request(&mut self, template: &RequestTemplate) -> HttpResult<Self::RequestType>;
+    /// Set a `ContextVerifier` callback that should be called if the request
+    /// is an HTTPS request.
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier);
+}
+
+/// A struct that represents the functionality of creating new HTTP/1.1
+/// requests.
+pub struct Http11RequestFactory {
+    /// The `Connector` instance used to establish a new connection to
+    /// servers.
+    connector: Connector,
+}
+
+impl Http11RequestFactory {
+    pub fn with_connector(c: Connector) -> Http11RequestFactory {
+        Http11RequestFactory {
+            connector: c,
+        }
+    }
+}
+
+impl HttpRequestFactory for Http11RequestFactory {
+    type RequestType = Request<Fresh>;
+
+    fn new() -> Http11RequestFactory {
+        Http11RequestFactory::with_connector(with_connector(HttpConnector(None)))
+    }
+
+    fn get_fresh_request(&mut self, template: &RequestTemplate) -> HttpResult<Request<Fresh>> {
+        let req = try!(Request::with_connector(
+                template.method.clone(),
+                template.url.clone(),
+                &mut self.connector));
+
+        Ok(req)
+    }
+
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier) {
+        self.connector = with_connector(HttpConnector(Some(verifier)));
+    }
+}
+
 /// A Client to use additional features with Requests.
 ///
 /// Clients can handle things such as: redirect policy.
-pub struct Client {
-    connector: Connector,
+pub struct Client<F=Http11RequestFactory> where F: HttpRequestFactory {
```

Comment:
Right, that's a nice hack and could be used to circumvent the problem I mentioned.

That said, I'm still not convinced of the benefit of doing it like that, instead of having a generic `Client`. Consider the `send` method on the `RequestBuilder`. In this scenario, its signature needs to change to return `HttpResult<Box<HttpResponse>>` -- since we can't know the actual type of the response. This is also a change to the public API of the client -- and in my opinion a more significant one than introducing a generic type parameter (with a default).

If the `send` method returns a trait object, all clients that previously did stuff like:

```rust
    // get res the usual way
    println!("Response: {}", res.status);
```

Would need to be changed to use the appropriate trait method, instead of directly accessing the field. This also implies having to import the trait itself.

Isn't this more invasive than adding a type annotation when declaring a client instance?

I have a branch [client-box-api](https://github.com/mlalic/hyper/tree/client-box-api) on my fork where you can see how it looks with this alternative approach. (It also includes the HTTP/2 commits)